File: actix/src/lib/scorecard.rs
use super::Dice;

pub struct Scorecard {
    ones: Option<u16>,
    twos: Option<u16>,
    threes: Option<u16>,
    fours: Option<u16>,
    fives: Option<u16>,
    sixes: Option<u16>,
    three_of_a_kind: Option<u16>,
    four_of_a_kind: Option<u16>,
    full_house: Option<u16>,
    small_straight: Option<u16>,
    large_straight: Option<u16>,
    yahtzee: Option<u16>,
    chance: Option<u16>,
}

impl Scorecard {
    pub fn new() -> Self {
        Self {
            ones: None,
            twos: None,
            threes: None,
            fours: None,
            fives: None,
            sixes: None,
            three_of_a_kind: None,
            four_of_a_kind: None,
            full_house: None,
            small_straight: None,
            large_straight: None,
            yahtzee: None,
            chance: None,
        }
    }

    pub fn score_upper(&mut self, val: &str, dice: Dice) {
        match val {
            "ones" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 1 {
                        total += 1;
                        self.ones = Some(total);
                    }
                }
            }
            "twos" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 2 {
                        total += 1;
                        self.twos = Some(total);
                    }
                }
            }
            "threes" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 3 {
                        total += 1;
                        self.threes = Some(total);
                    }
                }
            }
            "fours" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 4 {
                        total += 1;
                        self.fours = Some(total);
                    }
                }
            }
            "fives" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 5 {
                        total += 1;
                        self.fives = Some(total);
                    }
                }
            }
            "sixes" => {
                let mut total: u16 = 0;
                for die in dice.iter() {
                    if die.val() == 6 {
                        total += 1;
                        self.sixes = Some(total);
                    }
                }
            }
            _ => {}
        }
    }

    fn validate_board(&self) -> bool {
        self.ones.is_some()
            && self.twos.is_some()
            && self.threes.is_some()
            && self.fours.is_some()
            && self.fives.is_some()
            && self.sixes.is_some()
            && self.three_of_a_kind.is_some()
            && self.four_of_a_kind.is_some()
            && self.full_house.is_some()
            && self.small_straight.is_some()
            && self.large_straight.is_some()
            && self.yahtzee.is_some()
            && self.chance.is_some()
    }

    pub fn get_total(&self) -> Result<u16, String> {
        if !self.validate_board() {
            return Err("Not all fields are filled out".to_string());
        }
        let mut total = 0;
        total += self.ones.unwrap();
        total += self.twos.unwrap();
        total += self.threes.unwrap();
        total += self.fours.unwrap();
        total += self.fives.unwrap();
        total += self.sixes.unwrap();
        total += self.three_of_a_kind.unwrap();
        total += self.four_of_a_kind.unwrap();
        total += self.full_house.unwrap();
        total += self.small_straight.unwrap();
        total += self.large_straight.unwrap();
        total += self.yahtzee.unwrap();
        total += self.chance.unwrap();

        Ok(total)
    }
}

File: ytz/lib/ytz/game/scorecard.ex
defmodule Ytz.Game.Scorecard do
  alias Ytz.Game.{Dice, Scoring}

  defstruct ones: nil,
            twos: nil,
            threes: nil,
            fours: nil,
            fives: nil,
            sixes: nil,
            three_of_a_kind: nil,
            four_of_a_kind: nil,
            full_house: nil,
            small_straight: nil,
            large_straight: nil,
            yahtzee: nil,
            chance: nil

  @type t :: %__MODULE__{
          ones: non_neg_integer() | nil,
          twos: non_neg_integer() | nil,
          threes: non_neg_integer() | nil,
          fours: non_neg_integer() | nil,
          fives: non_neg_integer() | nil,
          sixes: non_neg_integer() | nil,
          three_of_a_kind: non_neg_integer() | nil,
          four_of_a_kind: non_neg_integer() | nil,
          full_house: non_neg_integer() | nil,
          small_straight: non_neg_integer() | nil,
          large_straight: non_neg_integer() | nil,
          yahtzee: non_neg_integer() | nil,
          chance: non_neg_integer() | nil
        }

  @categories [
    :ones,
    :twos,
    :threes,
    :fours,
    :fives,
    :sixes,
    :three_of_a_kind,
    :four_of_a_kind,
    :full_house,
    :small_straight,
    :large_straight,
    :yahtzee,
    :chance
  ]

  @typedoc "A valid scorecard category (keys of the scorecard struct)"
  @type category :: atom()

  defguard valid_category?(category)
           when is_atom(category) and category in @categories

  def new do
    %__MODULE__{}
  end

  def available_categories(%__MODULE__{} = scorecard) do
    scorecard
    |> Map.from_struct()
    |> Enum.filter(fn {_category, points} -> points == nil end)
    |> Enum.map(fn {category, _points} -> category end)
  end

  def available_categories(_scorecard) do
    {:error, "Invalid scorecard provided"}
  end

  def available_categories(%__MODULE__{} = scorecard, %Dice{} = dice) do
    scorecard
    |> Map.from_struct()
    |> Enum.filter(fn {_category, points} -> points == nil end)
    |> Enum.filter(fn {category, _points} -> Scoring.valid_for_category?(dice, category) end)
    |> Enum.map(fn {category, _points} -> category end)
  end

  def available_categories(_scorecard, %Dice{} = _dice) do
    {:error, "Invalid scorecard provided"}
  end

  def available_categories(%__MODULE__{} = _scorecard, _dice) do
    {:error, "Invalid dice provided"}
  end

  def available_categories(_scorecard, _dice) do
    {:error, "Invalid scorecard and dice provided"}
  end

  def categories, do: @categories

  def upsert_score(scorecard, _category, _points) when not is_struct(scorecard, __MODULE__) do
    {:error, "Invalid scorecard provided"}
  end

  def upsert_score(%__MODULE__{} = _scorecard, category, _points)
      when not valid_category?(category) do
    {:error, "Invalid category"}
  end

  def upsert_score(_scorecard, _category, points) when points < 0 do
    {:error, "Points cannot be negative"}
  end

  def upsert_score(%__MODULE__{} = scorecard, category, points) when valid_category?(category) do
    Map.put(scorecard, category, points)
  end

  # TODO: move this function to the scoring module
  def calculate_score(category, %Dice{} = dice) do
    case category do
      :ones -> Scoring.sum_dice_values(dice, 1)
      :twos -> Scoring.sum_dice_values(dice, 2)
      :threes -> Scoring.sum_dice_values(dice, 3)
      :fours -> Scoring.sum_dice_values(dice, 4)
      :fives -> Scoring.sum_dice_values(dice, 5)
      :sixes -> Scoring.sum_dice_values(dice, 6)
      :three_of_a_kind -> Scoring.calculate_three_of_a_kind(dice)
      :four_of_a_kind -> Scoring.calculate_four_of_a_kind(dice)
      :full_house -> Scoring.calculate_full_house(dice)
      :small_straight -> Scoring.calculate_small_straight(dice)
      :large_straight -> Scoring.calculate_large_straight(dice)
      :yahtzee -> Scoring.calculate_yahtzee(dice)
      :chance -> Scoring.sum_dice_values(dice)
      _ -> {:error, "Invalid category"}
    end
  end

  def calculate_score(_category, _dice) do
    {:error, "Invalid dice provided"}
  end

  def calculate_score(scorecard, category, %Dice{} = dice) do
    scorecard
    |> available_categories(dice)
    |> case do
      {:error, _} = error ->
        error

      available ->
        if category in available do
          calculate_score(category, dice)
        else
          {:error, "Category already filled or invalid for current dice"}
        end
    end
  end

  # TODO: implement catefory_filled?/2 and test so I can then use it to simplify calculate_score/3
  # def category_filled?(scorecard, category) do
  # def upper_total(scorecard) do
  # def upper_bonus(scorecard) do -> 35 or 0 based on upper total
  # def lower_total(scorecard) do
  # def total_score(scorecard) do
end
